documentation/developer_setup.md

```diff
@@ -0,0 +1,126 @@
+
+### Defining Methods
+Defining a method follows the format:
+```golang
+func (m *<MyStruct>) MethodName(param1 param1Type, param2 param2Type) (ReturnType, ReturnType2) {
+
+}
+```
+The `(m *<MyStruct>)` portion is the "receiver". Methods can be defined for either pointer (with a `*`) or receiver (without a `*`) types. Pointer receivers will avoid copying types on method calls and allow the method to mutate the receiving struct. You should use pointer receivers wherever possible to limit memory copies.
+
+
+Both public and private methods can be declared on clients. Below is an example in the `aztables` package for a `Create` method on the `TableServiceClient`:
+```golang
+// Create creates a table with the specified name.
+func (t *TableServiceClient) Create(ctx context.Context, name string) (TableCreateResponse, error) {
+	resp, err := t.client.Create(ctx, TableProperties{&name}, new(TableCreateOptions), new(QueryOptions))
+	if err == nil {
+		tableResp := resp.(TableCreateResponse)
+		return tableResp, nil
+	}
+	return TableCreateResponse{}, err
+}
+```
+
+All methods that perform I/O of any kind, sleep, or perform a significant amount of CPU-bound work must have the first parameter be of type [`context.Context`][golang_context] which allows the customer to carry a deadline, cancellation signal, and other values across API boundaries. The remaining parameters should be parameters specific to that method. The return types for methods should be first a "Response" object and second an `error` object.
+
+## Write Tests
+
+Testing is built into the Go toolchain as well with the `testing` library. The testing infrastructure located in the `sdk/internal` directory takes care of generating recordings, establishing the mode a test is being run in (options are "recording", "playback", "live-no-playback"), and reading environment variables.
+
+A simple test for `aztables` is shown below:
+```golang
+
+import (
+	"os"
+
+	"github.com/testify/assert"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/recording"
+)
+
+const (
+	accountName := os.GetEnv("TABLES_PRIMARY_ACCOUNT_NAME")
+	accountKey := os.GetEnv("TABLES_PRIMARY_ACCOUNT_KEY")
+	mode := recording.Recording
+)
+
+// Test creating a single table
+func TestCreateTable(t *testing.T) {
+	client := NewTableClient(accountName, accountKey, "tableName")
+	resp, err := client.Create()
+	assert.Nil(t, err)
```

Comment:
I'm not sure if it's better but there's also `assert.NoError(t, err)`. The output is maybe nicer since it's geared towards errors but I don't think it's a huge deal.

```
D:\Dev\temp\go1>go test
--- FAIL: TestThis (0.00s)
    main_test.go:11:
                Error Trace:    main_test.go:11
                Error:          Received unexpected error:
                                this is an error
                Test:           TestThis
FAIL
exit status 1
FAIL    whatever        0.186s

D:\Dev\temp\go1>go test
--- FAIL: TestThis (0.00s)
    main_test.go:12:
                Error Trace:    main_test.go:12
                Error:          Expected nil, but got: &errors.errorString{s:"this is an error"}
                Test:           TestThis
FAIL
exit status 1
FAIL    whatever        0.182s
```

Reply:
Thanks for pointing this out, this is way nicer on format. Also, I have been using `require` instead of `assert` because require will fail immediately rather than assert which will continue on in the test.